packages/hardhat-ledger/src/provider.ts

```diff
@@ -0,0 +1,445 @@
+import { ethers } from "ethers";
+import * as t from "io-ts";
+
+import { isValidAddress } from "@nomicfoundation/ethereumjs-util";
+
+import { isEIP712Message, ledgerService } from "@ledgerhq/hw-app-eth";
+import TransportNodeHid from "@ledgerhq/hw-transport-node-hid";
+import { EIP712Message } from "@ledgerhq/hw-app-eth/lib/modules/EIP712";
+import { EIP1193Provider, RequestArguments } from "hardhat/types";
+import { validateParams } from "hardhat/internal/core/jsonrpc/types/input/validation";
+import { rpcTransactionRequest } from "hardhat/internal/core/jsonrpc/types/input/transactionRequest";
+import {
+  rpcAddress,
+  rpcData,
+  rpcQuantityToBigInt,
+} from "hardhat/internal/core/jsonrpc/types/base-types";
+import { ProviderWrapperWithChainId } from "hardhat/internal/core/providers/chainId";
+import { HardhatError } from "hardhat/internal/core/errors";
+import { ERRORS } from "hardhat/internal/core/errors-list";
+
+import { wrapTransport } from "./internal/wrap-transport";
+import * as cache from "./internal/cache";
+import { LedgerOptions, EthWrapper, Signature, Paths } from "./types";
+import {
+  HardhatLedgerConnectionError,
+  HardhatLedgerDerivationPathError,
+  HardhatLedgerError,
+  HardhatLedgerNotControlledAddressError,
+} from "./errors";
+
+export class LedgerProvider extends ProviderWrapperWithChainId {
+  public static readonly MAX_DERIVATION_ACCOUNTS = 20;
+  public static readonly DEFAULT_TIMEOUT = 3000;
+
+  public readonly paths: Paths = {}; // { address: path }
+  public name: string = "LedgerProvider";
+
+  protected _eth: EthWrapper | undefined;
+  private _isCreatingTransport = false;
+
+  public static async create(
+    options: LedgerOptions,
+    _wrappedProvider: EIP1193Provider
+  ) {
+    const provider = new LedgerProvider(options, _wrappedProvider);
+    await provider.init();
+    return provider;
+  }
+
+  constructor(
+    public readonly options: LedgerOptions,
+    protected readonly _wrappedProvider: EIP1193Provider
+  ) {
+    super(_wrappedProvider);
+
+    this.options.accounts = options.accounts.map((account) => {
+      if (!isValidAddress(account)) {
+        throw new HardhatLedgerError(
+          `The following ledger address from the config is invalid: ${account}`
+        );
+      }
+      return account.toLowerCase();
+    });
+  }
+
+  public get eth(): EthWrapper {
+    if (this._eth === undefined) {
+      throw new HardhatError(ERRORS.GENERAL.UNINITIALIZED_PROVIDER);
+    }
+    return this._eth;
+  }
+
+  public async init(): Promise<void> {
+    if (this._eth === undefined && !this._isCreatingTransport) {
```

Comment:
What do you think about removing it and leaving a comment explaining the situation (and maybe linking to this comment)?
Just because I think we can easily forget about it and in case someone is confused by an error the lib is throwing and decides to read the code.

(I'll [just remove it](https://github.com/NomicFoundation/hardhat/pull/4008/commits/ceb7dfa5911533512926fb283219d7b3be1dc962) for now and add the comment later if you think it's a good idea)

Reply:
Makes sense, let's do that.